components/board/genericlinux/board.go

```
@@ -86,43 +76,127 @@ func (b *Board) Reconfigure(
+				return err
+			}
+			delete(b.gpios, oldName)
+			continue
+		}
+
+		// If we get here, the old pin definition exists, but the old pin does not. Check if it's a
+		// digital interrupt.
+		if interrupt, ok := b.interrupts[oldName]; ok {
+			if err := interrupt.Close(); err != nil {
+				return err
+			}
+			delete(b.interrupts, oldName)
+			continue
+		}
+
+		b.logger.Warnf("During reconfiguration, old pin '%s' should be destroyed, but "+
+			"it doesn't exist!?", oldName)
 	}
+
+	// Next, compare the new pin definitions to the old ones, to build up 2 sets: pins to rename,
+	// and new pins to create. Don't actually create any yet, in case you'd overwrite a pin that
+	// should be renamed out of the way first.
+	toRename := map[string]string{} // Maps old names for pins to new names
+	toCreate := map[string]GPIOBoardMapping{}
+	for newName, mapping := range newConf.GpioMappings {
+		if oldName, ok := getMatchingPin(mapping, b.gpioMappings); ok {
+			if oldName != newName {
+				toRename[oldName] = newName
+			}
+		} else {
+			toCreate[newName] = mapping
+		}
+	}
+
+	// Rename the ones whose name changed. The ordering here is tricky: if B should be renamed to C
+	// while A should be renamed to B, we need to make sure we don't overwrite B with A and then
+	// rename it to C. To avoid this, move all the pins to rename into a temporary data structure,
+	// then move them all back again afterward.
+	tempGpios := map[string]*gpioPin{}
+	tempInterrupts := map[string]*digitalInterrupt{}
+	for oldName, newName := range toRename {
+		if pin, ok := b.gpios[oldName]; ok {
+			tempGpios[newName] = pin
+			delete(b.gpios, oldName)
+			continue
+		}
+
+		// If we get here, again check if the missing pin is a digital interrupt.
+		if interrupt, ok := b.interrupts[oldName]; ok {
+			tempInterrupts[newName] = interrupt
+			delete(b.interrupts, oldName)
+			continue
+		}
+
+		b.logger.Errorf("During reconfiguration, old pin '%s' should be renamed to '%s', but "+
```

Maybe this should return an error instead of merely logging one?

---

Would there be a scenario where you would expect this error to happen?

---

Oh, snap! That was the perfect question to ask. No, I'd never expect this to happen. Therefore, we should return an error, because if we encounter this, it means there's probably a bug in our logic somewhere. Thank you for the insight!